pkg/config/remote/service/service.go

```diff
@@ -0,0 +1,535 @@
+				checksum = sha512Checksum[:]
+			default:
+				return fmt.Errorf("unsupported checksum %s", algorithm)
+			}
+
+			if !bytes.Equal(checksum, targetMeta.Hashes[algorithm]) {
+				return fmt.Errorf("target file %s has invalid checksum", string(checksum))
+			}
+		}
+	}
+
+	return nil
+}
+
+type bufferDestination struct {
+	bytes.Buffer
+}
+
+func (d *bufferDestination) Delete() error {
+	return nil
+}
+
+func (s *Service) notifySubscriber(subscriber *Subscriber, configResponse *pbgo.ConfigResponse) error {
+	log.Debugf("Notifying subscriber %s with version %d", subscriber.product, configResponse.DirectoryTargets.Version)
+
+	if err := subscriber.callback(configResponse); err != nil {
+		return err
+	}
+
+	subscriber.lastUpdate = time.Now()
+	subscriber.lastVersion = configResponse.DirectoryTargets.Version
+
+	return nil
+}
+
+// RegisterSubscriber registers a new subscriber for a product's configurations
+func (s *Service) RegisterSubscriber(subscriber *Subscriber) {
+	s.Lock()
+	s.subscribers = append(s.subscribers, subscriber)
+	s.Unlock()
+
+	product := subscriber.product
+	log.Debugf("New registered subscriber for %s", product.String())
+
+	config, err := s.store.GetLastConfig(product.String())
+	if err == nil {
+		log.Debugf("Found cached configuration for product %s", product)
+		if err := s.notifySubscriber(subscriber, config); err != nil {
+			log.Error(err)
+		}
+	} else {
+		log.Debugf("No stored configuration for product %s", product)
+	}
+}
+
+// UnregisterSubscriber unregisters a subscriber for a product's configurations
+func (s *Service) UnregisterSubscriber(unregister *Subscriber) {
+	s.Lock()
+	for i, subscriber := range s.subscribers {
+		if subscriber == unregister {
+			s.subscribers = append(s.subscribers[:i], s.subscribers[i+1:]...)
+		}
+	}
+	s.Unlock()
+}
+
+// Start the remote configuration management service
+func (s *Service) Start(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	ticker := time.NewTicker(s.opts.RefreshInterval)
+
+	go func() {
+		defer ticker.Stop()
+		defer cancel()
+
+		for {
+			select {
+			case <-ticker.C:
+				s.refresh()
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// GetConfigs returns config for the given product
+func (s *Service) GetConfigs(product string) ([]*pbgo.ConfigResponse, error) {
+	return s.store.GetConfigs(product)
+}
+
+// GetStore returns the configuration store
+func (s *Service) GetStore() *store.Store {
+	return s.store
+}
+
+// NewService instantiates a new remote configuration management service
+func NewService(opts Opts) (*Service, error) {
+	if opts.RefreshInterval <= 0 {
+		opts.RefreshInterval = config.Datadog.GetDuration("remote_configuration_refresh_interval")
+	}
+
+	if opts.RefreshInterval < minimalRefreshInterval {
+		opts.RefreshInterval = minimalRefreshInterval
+	}
+
+	if opts.DBPath == "" {
+		opts.DBPath = path.Join(config.GetDefaultRunPath(), "remote-config.db")
```

Comment:
My understand was that it is already the case as `GetDefaultRunPath()` returns `defaultRunPath`, and in `pkg/config/config.go` there is a `config.BindEnvAndSetDefault("run_path", defaultRunPath)`

Reply:
Are you saying `defaultRunPath` will change its value based on configuration? I don't think it will, `defaultRunPath` s a const, and in the call you quote it simply provides default value for the config option. Actual value needs to be queried by `config.Datadog.GetString()` or similar.

Reply:
I read the code too fast. I mistakenly thought that it was binding to the content of defaultRunPath.
I changed the code of GetDefaultRunPath() to take `run_path` into account